File: web/src/sim/world/NationStore.ts
export enum Doctrine {
  Western = 0,
  Eastern = 1,
  European = 2,
}

export class NationStore {
  readonly count: number

  readonly tag: string[]
  readonly name: string[]
  readonly doctrine: Uint8Array
  readonly capitalProvince: Int32Array
  readonly isAlive: Uint8Array

  constructor(count: number) {
    if (count < 0) {
      throw new RangeError(`jumlah negara tidak boleh negatif: ${count}`)
    }

    this.count = count
    this.tag = new Array<string>(count).fill('')
    this.name = new Array<string>(count).fill('')
    this.doctrine = new Uint8Array(count)
    this.capitalProvince = new Int32Array(count).fill(-1)
    this.isAlive = new Uint8Array(count).fill(1)
  }

  indexOf(tag: string): number {
    return this.tag.indexOf(tag)
  }
}

File: sim/NationRise.Core/World/NationStore.cs
namespace NationRise.Core.World;

public enum Doctrine : byte
{
    Western = 0,
    Eastern = 1,
    European = 2,
}

public sealed class NationStore
{
    public int Count { get; }

    public string[] Tag { get; }
    public string[] Name { get; }
    public Doctrine[] Doctrine { get; }
    public int[] CapitalProvince { get; }
    public bool[] IsAlive { get; }

    public NationStore(int count)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(count);

        Count = count;
        Tag = new string[count];
        Name = new string[count];
        Doctrine = new Doctrine[count];
        CapitalProvince = new int[count];
        IsAlive = new bool[count];

        Array.Fill(Tag, string.Empty);
        Array.Fill(Name, string.Empty);
        Array.Fill(CapitalProvince, -1);
        Array.Fill(IsAlive, true);
    }

    public int IndexOf(string tag)
    {
        for (int i = 0; i < Count; i++)
        {
            if (Tag[i] == tag)
            {
                return i;
            }
        }

        return -1;
    }
}
